File: HourRank_30_P1.py
# code_report Solution
# https://youtu.be/uOG3QyxIjso

# Solution 1:

import defaultdict from collections

def solve(names):
    s = set()
    d = defaultdict(int)
    l = list()
    for name in names:
        if name in d:
            d[name] += 1
            l.append(name+" "+str(d[name]))
        else:
            d[name] = 1
            t = ""
            inserted = False;
            for i in range(len(name)):
                t += name[i:i+1]
                if t not in s and not inserted:
                    inserted = True
                    l.append(t)
                s.add(t)
            if not inserted: 
                l.append(name)
    return l

#Solution 2: 

def solve(names):
    trie = {}
    res = []
    for name in names:
        node = trie
        printed = False
        for i, x in enumerate(name, start=1):
            if x not in node:
                if not printed: 
                    res.append(name[:i])
                    printed = True
                node[x] = defaultdict(int)

            node = node[x]

        node['count'] += 1

        if not printed: 
            res.append (name + (' '+str(node['count']) if node['count']!=1 else ''))
    return res

File: HourRank_30_P1.cpp
// code_report Solution

// Solution 1:
// https://youtu.be/uOG3QyxIjso

vector<string> solve (vector<string> names) {
   unordered_set<string> s;
   unordered_map<string, int> m;
   vector<string> ans;
   for (const auto& name : names) {
      auto it = m.find (name);
      if (it != m.end ()) {
         it->second++;
         ans.push_back (name + ' ' + to_string (it->second));
      }
      else {
         m[name] = 1;
         string t;
         bool inserted = false;
         for (auto c : name) {
            t += c;
            auto p = s.insert (t);
            if (!inserted && p.second) {
               inserted = true;
               ans.push_back (t);
            }
         }
         if (!inserted) ans.push_back (t); // very tricky case
      }
   }
   return ans;
}

// Solution 2 (Trie):
// https://youtu.be/VsiP-dTWyG4?t=249

struct trie { 
   unordered_map<char, trie*> m; 
   int count = 0;
};

vector<string> solve(vector<string> names) {
   auto* t = new trie();
   vector<string> res;
   for (const auto& name : names) {
      auto* node = t;
      auto added = false;
      auto p = ""s;
      for (auto c : name) {
         p += c; // prefix
         if (!node->m.count(c)) {
            if (!added) res.push_back(p), added = true;
            node->m[c] = new trie();
         }
         node = node->m[c];
      }
      node->count++;

      if (!added)
         res.push_back(p + (node->count != 1 ? " " + to_string(node->count) : ""));
   }
   return res;
}
